pcmaudiotrack.go

```diff
@@ -0,0 +1,192 @@
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/gammazero/deque"
+	audio "github.com/livekit/mediatransportutil/pkg/audio"
+	opus "github.com/livekit/mediatransportutil/pkg/audio/opus"
+	rtp "github.com/livekit/mediatransportutil/pkg/audio/rtp"
+	protoLogger "github.com/livekit/protocol/logger"
+	"github.com/pion/webrtc/v4"
+	"go.uber.org/atomic"
+)
+
+const (
+	opusSampleRate = 48000
+)
+
+type PCM16ToOpusAudioTrack struct {
+	*webrtc.TrackLocalStaticSample
+
+	opusWriter         audio.WriteCloser[opus.Sample]
+	pcmWriter          audio.WriteCloser[audio.PCM16Sample]
+	resampledPCMWriter audio.WriteCloser[audio.PCM16Sample]
+
+	frameDuration time.Duration
+
+	sampleBuffer *deque.Deque[audio.PCM16Sample]
+	started      sync.Once
+	ticker       *time.Ticker
+
+	closed atomic.Bool
+	mu     sync.Mutex
+	cond   *sync.Cond
+}
+
+// TODO: Support stereo
+func NewPCM16ToOpusAudioTrack(sampleRate int, frameDuration time.Duration, logger protoLogger.Logger) (*PCM16ToOpusAudioTrack, error) {
+	track, err := webrtc.NewTrackLocalStaticSample(webrtc.RTPCodecCapability{MimeType: webrtc.MimeTypeOpus}, "test", "test")
+	if err != nil {
+		return nil, err
+	}
+
+	opusWriter := audio.FromSampleWriter[opus.Sample](track, opusSampleRate, frameDuration)
+	pcmWriter, err := opus.Encode(opusWriter, 1, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	resampledPCMWriter := pcmWriter
+	if sampleRate != opusSampleRate {
+		resampledPCMWriter = audio.ResampleWriter(pcmWriter, opusSampleRate)
+	}
+
+	t := &PCM16ToOpusAudioTrack{
+		TrackLocalStaticSample: track,
+		opusWriter:             opusWriter,
+		pcmWriter:              pcmWriter,
+		resampledPCMWriter:     resampledPCMWriter,
+		frameDuration:          frameDuration,
+		sampleBuffer:           new(deque.Deque[audio.PCM16Sample]),
+	}
+	t.cond = sync.NewCond(&t.mu)
+
+	go t.processSamples()
+	return t, nil
+}
+
+func (t *PCM16ToOpusAudioTrack) WriteSample(sample audio.PCM16Sample) error {
+	if t.closed.Load() {
+		return errors.New("track is closed")
+	}
+
+	t.mu.Lock()
+	t.sampleBuffer.PushBack(sample)
+	t.cond.Broadcast()
+	t.mu.Unlock()
+	return nil
+}
+
+func (t *PCM16ToOpusAudioTrack) waitForSamples() {
+	t.mu.Lock()
+	for t.sampleBuffer.Len() == 0 && !t.closed.Load() {
+		t.cond.Wait()
+	}
+	t.mu.Unlock()
+}
+
+func (t *PCM16ToOpusAudioTrack) processSamples() {
+	// wait for the first sample before starting the ticker
+	t.waitForSamples()
+	t.resampledPCMWriter.WriteSample(t.sampleBuffer.PopFront())
+	t.ticker = time.NewTicker(t.frameDuration)
+
+	for range t.ticker.C {
+		isBufferEmpty := t.sampleBuffer.Len() == 0
+		if t.closed.Load() {
+			return
+		}
+		if isBufferEmpty {
+			t.waitForSamples()
+		}
+
+		t.mu.Lock()
+		sample := t.sampleBuffer.PopFront()
+
+		if isBufferEmpty {
+			t.ticker.Reset(t.frameDuration)
+		}
+
+		t.resampledPCMWriter.WriteSample(sample)
+		t.mu.Unlock()
+	}
+
+	t.sampleBuffer.Clear()
+	t.ticker.Stop()
+}
+
+func (t *PCM16ToOpusAudioTrack) Close() {
+	t.closed.Store(true)
```

Comment:
Should probably be `CompareAndSwap`, if it fails then `Close` should just return. This protect from double-close, which usually happens when doing `defer Close()` and manually closing somewhere else.

Reply:
done